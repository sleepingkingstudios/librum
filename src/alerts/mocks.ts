import type {
  Alert,
  DismissAlert,
  DismissAllAlerts,
  DisplayAlert,
} from '@alerts';

type MockAlertsContext = {
  alerts: Alert[],
  dismissAlert: jest.MockedFunction<DismissAlert>,
  dismissAllAlerts: jest.MockedFunction<DismissAllAlerts>,
  displayAlert: jest.MockedFunction<DisplayAlert>,
};

type MockUseAlerts = jest.MockedFunction<() => MockAlertsContext>;

const alerts: MockAlertsContext = {
  alerts: [] as Alert[],
  dismissAlert: jest.fn(),
  dismissAllAlerts: jest.fn(),
  displayAlert: jest.fn(),
};

export const useAlerts: MockUseAlerts = jest.fn(() => alerts);
