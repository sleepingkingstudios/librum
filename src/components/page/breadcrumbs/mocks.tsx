import * as React from 'react';

import type { Breadcrumb } from './types';

const formatBreadcrumb = (breadcrumb: Breadcrumb): string => {
  const { active, label, url } = breadcrumb;

  return `${label}${url ? ` @ ${url}` : ''}${active ? ' (active)' : ''}`;
};

export const PageBreadcrumbs = ({
  breadcrumbs,
}: {
  breadcrumbs: Breadcrumb[],
}): JSX.Element => (
  <ul aria-label="breadcrumbs">
    {
      breadcrumbs.map((breadcrumb: Breadcrumb) => (
        <li aria-label="breadcrumb" key={breadcrumb.label}>
          { formatBreadcrumb(breadcrumb) }
        </li>
      ))
    }
  </ul>
);
