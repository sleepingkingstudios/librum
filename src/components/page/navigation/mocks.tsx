import * as React from 'react';

import type { DropdownProps } from './dropdown';
import type { NavigationItemProps } from './item';

const renderNavigationItem = (
  item: DropdownProps | NavigationItemProps
): JSX.Element => {
  const { label } = item;

  if ('items' in item) {
    return (
      <li key={label}>
        { label }
        <ul>
          { item.items.map(renderNavigationItem) }
        </ul>
      </li>
    );
  }

  const { url } = item;

  return (
    <li aria-label="navigation-item" key={label}>{ `${label} @ ${url}` }</li>
  );
};

export const PageNavigation = ({
  navigation,
}: {
  navigation: Array<DropdownProps | NavigationItemProps>,
}): JSX.Element => {
  if (!navigation) { return null; }

  if (navigation.length === 0) { return null; }

  return (
    <ul aria-label="navigation">
      {
        navigation.map(renderNavigationItem)
      }
    </ul>
  );
};
