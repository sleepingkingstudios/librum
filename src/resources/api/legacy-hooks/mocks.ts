import { defaultResponse } from '@api/test-helpers';
import type { UseResourceQuery } from '../types';

const resourceQuery: jest.MockedFunction<UseResourceQuery> =
  jest.fn(() => defaultResponse);

export const useResourceQuery: jest.MockedFunction<() => UseResourceQuery> =
  jest.fn(() => resourceQuery);
