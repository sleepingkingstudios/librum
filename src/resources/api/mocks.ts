import type { UseQuery } from '@api';
import { defaultResult } from '@api/test-helpers';
import type {
  ResourceApiEndpointConfiguration,
  ResourceConfiguration,
} from '../types';
import type { ResourceApiHooks } from './types';

const useIndexResources: jest.MockedFunction<UseQuery> =
  jest.fn(() => defaultResult);

type GenerateResourcesApiProps = {
  endpoints?: ResourceApiEndpointConfiguration,
} & ResourceConfiguration;

type GenerateResourcesApi =
  (options: GenerateResourcesApiProps) => ResourceApiHooks;

export const generateResourcesApi: jest.MockedFunction<GenerateResourcesApi> =
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  jest.fn((...args) => ({
    useIndexResources,
  }));
