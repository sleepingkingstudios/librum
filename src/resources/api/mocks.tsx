import { responseWithStatus } from '@api';
import type {
  Refetch,
  Response,
} from '@api';

type UseResourceQuery = (props: unknown) => [Response, Refetch];

export { generateAlerts } from './utils';

export const useResourceQuery: jest.MockedFunction<UseResourceQuery> = jest.fn(
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  (props) => [responseWithStatus({ status: 'uninitialized' }), jest.fn()]
);
