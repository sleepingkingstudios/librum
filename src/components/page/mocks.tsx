import * as React from 'react';

import type {
  Breadcrumb,
  NavigationProps,
} from './index';
import { isElement } from '@utils/react-utils';
import { PageBreadcrumbs } from './breadcrumbs/mocks';
import { PageNavigation } from './navigation/mocks';

const renderBreadcrumbs = ({
  breadcrumbs,
}: {
  breadcrumbs?: JSX.Element | Breadcrumb[],
}): JSX.Element => {
  if (!breadcrumbs) { return null; }

  if (isElement(breadcrumbs)) { return breadcrumbs; }

  const configured = breadcrumbs as Breadcrumb[];

  if (configured.length === 0) { return null; }

  return (<PageBreadcrumbs breadcrumbs={configured} />);
};

const renderNavigation = ({
  navigation,
}: {
  navigation: JSX.Element | NavigationProps,
}): JSX.Element => {
  if (!navigation) { return null; }

  if (!('length' in navigation)) { return navigation; }

  if (navigation.length === 0) { return null; }

  return (<PageNavigation navigation={navigation} />);
};

const renderTitle = ({
  subtitle,
  title,
}: {
  subtitle?: string,
  title?: string,
}): JSX.Element => {
  if (!title || title.length === 0) { return null; }

  if (!subtitle || subtitle.length === 0) {
    return (<span aria-label="title">{ title }</span>);
  }

  return (<span aria-label="title">{ `${title}: ${subtitle}` }</span>);
};

export const Page = ({
  breadcrumbs,
  children,
  navigation,
  subtitle,
  title,
}: {
  breadcrumbs?: JSX.Element | Breadcrumb[],
  children: React.ReactNode,
  navigation?: JSX.Element | NavigationProps,
  subtitle?: string,
  title?: string,
}): JSX.Element => {
  return (
    <div className="mock-page">
      { renderTitle({ subtitle, title }) }

      { renderNavigation({ navigation }) }

      { children }

      { renderBreadcrumbs({ breadcrumbs }) }
    </div>
  );
};
