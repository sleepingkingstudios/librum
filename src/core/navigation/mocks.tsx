import * as React from 'react';

import { PageNavigation } from '@components/page/navigation/mocks';

const navigation = [
  {
    label: 'Nav Item',
    url: '/nav-item',
  },
  {
    label: 'Dropdown',
    items: [
      {
        label: 'Dropdown Item',
        url: '/dropdown-item',
      },
    ],
  },
];

export const CoreNavigation =
  (): JSX.Element => (<PageNavigation navigation={navigation} />);
