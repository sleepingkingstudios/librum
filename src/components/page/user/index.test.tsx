import * as React from 'react';
import { faUserXmark } from '@fortawesome/free-solid-svg-icons';

import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { PageUser } from './index';
import { useAlerts as mockUseAlerts } from '@alerts/mocks';
import {
  actions,
  selector,
} from '@session';
import type {
  Session,
  User,
} from '@session';
import { createStore } from '@test-helpers/store';

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@alerts', () => require('@alerts/mocks'));

const { displayAlert } = mockUseAlerts();

describe('<PageUser>', () => {
  beforeEach(() => { displayAlert.mockClear(); });

  it('should not show a user name', () => {
    const { queryByText } = render(<PageUser />, { store: true });

    expect(queryByText(/currently logged in/)).toBeNull();
  });

  it('should not show the logout button', () => {
    const { queryByRole } = render(<PageUser />, { store: true });

    expect(queryByRole('button', { name: 'Log Out' })).toBeNull();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(<PageUser />, { store: true });

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the session is authenticated', () => {
    const user: User = {
      email: 'alan.bradley@example.com',
      id: '00000000-0000-0000-0000-000000000000',
      role: 'user',
      slug: 'alan-bradley',
      username: 'Alan Bradley',
    };
    const token = '12345';
    const { create } = actions;

    it('should show the user name link', () => {
      const { dispatch, store } = createStore();
      dispatch(create({ token, user }));

      const { getByRole, getByText } = render(
        <PageUser />,
        {
          router: true,
          store,
        },
      );

      const notice = getByText(/currently logged in/);

      expect(notice).toBeVisible();
      expect(notice).toHaveTextContent('You are currently logged in as');

      expect(getByRole('link', { name: user.username })).toBeVisible();
    });

    it('should show the logout link', () => {
      const { dispatch, store } = createStore();
      dispatch(create({ token, user }));

      const { getByRole } = render(
        <PageUser />,
        {
          router: true,
          store,
        },
      );

      expect(getByRole('button', { name: 'Log Out' })).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { dispatch, store } = createStore();
      dispatch(create({ token, user }));

      const { asFragment } = render(
        <PageUser />,
        {
          router: true,
          store,
        },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the user clicks the logout button', () => {
      afterEach(() => { localStorage.clear(); });

      it('should not show a user name', async () => {
        const { dispatch, store } = createStore();
        dispatch(create({ token, user }));

        const { getByRole, queryByText } = render(
          <PageUser />,
          {
            router: true,
            store,
          },
        );
        const button = getByRole('button', { name: 'Log Out' });

        await userEvent.click(button);

        expect(queryByText(/currently logged in/)).toBeNull();
      });

      it('should display an alert', async () => {
        const expected = {
          context: 'authentication:session',
          icon: faUserXmark,
          message: 'You have successfully logged out.',
          type: 'warning',
        };
        const { dispatch, store } = createStore();

        dispatch(create({ token, user }));

        const { getByRole } = render(
          <PageUser />,
          {
            router: true,
            store,
          },
        );
        const button = getByRole('button', { name: 'Log Out' });

        await userEvent.click(button);

        expect(displayAlert).toHaveBeenCalledWith(expected);
      });

      it('should update the store', async () => {
        const expected = { authenticated: false };

        const { dispatch, getState, store } = createStore();
        dispatch(create({ token, user }));

        const { getByRole } = render(
          <PageUser />,
          {
            router: true,
            store,
          },
        );
        const button = getByRole('button', { name: 'Log Out' });

        await userEvent.click(button);

        const state = getState() as { session: Session };

        expect(selector(state)).toEqual(expected);
      });

      it('should clear the stored session', async () => {
        localStorage.setItem('session', JSON.stringify({ authenticated: true }));

        const { dispatch, store } = createStore();
        dispatch(create({ token, user }));

        const { getByRole } = render(
          <PageUser />,
          {
            router: true,
            store,
          },
        );
        const button = getByRole('button', { name: 'Log Out' });

        await userEvent.click(button);

        expect(localStorage.getItem('session')).toBeNull();
      });
    });
  });
});
