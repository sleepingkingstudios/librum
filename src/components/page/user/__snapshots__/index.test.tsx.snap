// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageUser> should match the snapshot 1`] = `<DocumentFragment />`;

exports[`<PageUser> when the session is authenticated should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="text-left mt-2"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-user mr-1"
      data-icon="user"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
        fill="currentColor"
      />
    </svg>
    <span
      class="hidden sm:inline"
    >
      You are currently logged in as 
    </span>
    <a
      href="/user"
    >
      Alan Bradley
    </a>
    <span
      class="float-right link-danger"
    >
      <button>
        Log Out
      </button>
    </span>
  </div>
</DocumentFragment>
`;
